Fix parseScanCursorOrReply to reject empty strings instead of treating them the same as "0"

Mixficsol opened the pull request:
Modified the judgment statement if in parseScanCursorOrReply so that an empty string is considered an error.
For empty string, strtoul returns 0, and eptr is set to the end of the string (`'\0'`), the existing code would have treated that as success.

Mixficsol commented:
Could you please Review this code for me @oranagra @guybe7

Mixficsol commented:
Could you please Review this code for me @madolson

oranagra commented:
@Mixficsol let me make sure i understand it correctly.
the bug is that an empty string is successfully parsed as 0, but there's no invalid memory access here, right? (since `eptr` is set to the input string, and does equal to `'\0'`.

btw, maybe the fix will look cleaner if we do `sdslen(o->ptr)==0` or `eptr==o->ptr`?

p.s. i see we have a similar problem in a few other places.

Mixficsol commented:
> @Mixficsol let me make sure i understand it correctly. the bug is that an empty string is successfully parsed as 0, but there's no invalid memory access here, right? (since `eptr` is set to the input string, and does equal to `'\0'`.
> 
> btw, maybe the fix will look cleaner if we do `sdslen(o->ptr)==0` or `eptr==o->ptr`?
> 
> p.s. i see we have a similar problem in a few other places.

Yes, I have submitted a version of the code again. Could you please help me merge the code after checking it? I will also continue to troubleshoot similar problems you mentioned in redis later

Mixficsol commented:
> @Mixficsol let me make sure i understand it correctly. the bug is that an empty string is successfully parsed as 0, but there's no invalid memory access here, right? (since `eptr` is set to the input string, and does equal to `'\0'`.
> 
> btw, maybe the fix will look cleaner if we do `sdslen(o->ptr)==0` or `eptr==o->ptr`?
> 
> p.s. i see we have a similar problem in a few other places.
<img width="562" alt="截屏2023-03-27 15 16 33" src="https://user-images.githubusercontent.com/73943232/227870685-61d8253f-a824-42c6-b06b-7dc3cb2bad6a.png">
<img width="587" alt="截屏2023-03-27 15 15 36" src="https://user-images.githubusercontent.com/73943232/227870708-551cb4be-8185-49f6-bfeb-1ce850c46c52.png">
This is the command output before and after the modification

Mixficsol commented:
@oranagra 这是修改前和修改后的命令输出
<img width="562" alt="截屏2023-03-27 15 16 33" src="https://user-images.githubusercontent.com/73943232/227871048-296077d7-afe4-4909-99ca-9c95273f7109.png">
<img width="587" alt="截屏2023-03-27 15 15 36" src="https://user-images.githubusercontent.com/73943232/227871079-4bd3dbbd-648b-42d4-ad67-399174336664.png">
I have submitted a version of the code again. Could you please help me merge the code after checking it? I will also continue to troubleshoot similar problems you mentioned in redis later

Mixficsol commented:
@oranagra This is the command output before and after the modification
<img width="562" alt="截屏2023-03-27 15 16 33" src="https://user-images.githubusercontent.com/73943232/227871280-d8a4a505-c9f7-46ae-92e3-8d51d2a5d143.png">
<img width="587" alt="截屏2023-03-27 15 15 36" src="https://user-images.githubusercontent.com/73943232/227871309-1b6dacef-ed97-4b23-b661-5d8cf338177e.png">

 I have submitted a version of the code again. Could you please help me merge the code after checking it? I will also continue to troubleshoot similar problems you mentioned in redis later

oranagra commented:
looks good to me (although `eptr == o->ptr` may be even nicer).
let's fix all similar problems in redis inside the same PR (and merge it when ready).
thanks.

hwware commented:
@Mixficsol I check our redis test suites, I think the scan command do not touch this edge case, could you please add this to our test?  If you find similar issue in our codes, could you add them as well?   Thanks

vitarb commented on src/db.c:
```diff
@@ -835,7 +835,7 @@ int parseScanCursorOrReply(client *c, robj *o, unsigned long *cursor) {
     errno = 0;
     *cursor = strtoul(o->ptr, &eptr, 10);
-    if (isspace(((char*)o->ptr)[0]) || eptr[0] != '\0' || errno == ERANGE)
+    if (sdslen(o->ptr) == 0 || isspace(((char*)o->ptr)[0]) || eptr[0] != '\0' || errno == ERANGE)
```
Should we instead just check that first char `!isdigit` instead of `isspace`?
e.g. `if (!isdigit(((char*)o->ptr)[0]) || eptr[0] != '\0' || errno == ERANGE)`

oranagra replied:
i think i'd rather have all the code that does similar things (strtoll, strtoul, strtod) to be somewhat similar (we have this bug in many places), so considering it's gonna work on signed values, maybe it's better not to assume anything on the characters at all?

vitarb replied:
What other, similar, places did you have in mind, maybe I just don't have right context?
I don't feel strongly about this, but the form I've proposed above seems to be most concise, yet readable way to express what we want. Cursor is unsigned, so it should be fine to check if first char is a digit.

oranagra replied:
i mean all the other places that call this family of APIs (strtoll, strtod) and use `&eptr` to validate they parsed the whole string.
e.g. there's one in setConfigOOMScoreAdjValuesOption, one in sortCommandGeneric, and many others.

Mixficsol replied:
@oranagra Hello, I would like you to just help me merge this PR, and I will continue to pay attention to all other places susceptible to this error in the future, and I will pay attention to those unexpected modifications (such as Spaces) in the future. Thank you again for helping me merge this PR, and I will definitely continue to pay attention to the follow-up problems of redis

oranagra replied:
i think i'd rather look at the wider problem in one go, rather than solving the issues one by one (which could lead to some realizations and then repeated changes to the same code).
and either way, i can't merge this PR in it's current form since it contains quite a few unintended changes.

enjoy-binbin commented on tests/unit/scan.tcl:
```diff
@@ -272,7 +277,7 @@ start_server {tags {"scan network"}} {
         lsort -unique [lindex $res 1]
     }
-
+    
```
```suggestion

```

enjoy-binbin commented on tests/unit/scan.tcl:
```diff
@@ -282,7 +287,7 @@ start_server {tags {"scan network"}} {
         assert {$first_score != 0}
     }
-
+    
```
```suggestion

```

enjoy-binbin reviewed (commented):
please these spaces

Mixficsol commented:
@oranagra Could you please help me merge the submitted code #11974, I have seen the problem with unstable branch these days, I haven't found the problem with sds unstable string yet, besides the one I changed before, I also added a test sample of this problem on Scans.tcl, could I also merge this with it if convenient? thank you

oranagra commented:
@Mixficsol i don't think i understand your last post, are you asking me for some guidance, or just that i merge this PR?
i see you added a test that looks fine, but also you made quite a few other unintended changes.

besides, i'd like to look into fixing all the other places that are vulnerable to this error (nearly everywhere that uses `&eptr` on `strtod`, `strtol` family. if you're not up to it, i can merge just the scan one and leave the others for later..
